File: scripts/get_module.py
import json
import os
import shutil
import sys
from urllib.parse import urlparse as parse_url
import utils

SERVER_DIR = utils.get_server_dir()

def get_dependency(url):
    parsed_url = parse_url("https://" + url)
    if parsed_url.netloc == "" or parsed_url.scheme == "":
        print("Please format dependency URLs as: username/repository (only Github repositories)")
        print("Couldn't get dependency: " + url)
        return
    print("Installing dependency... ({})".format(url))
    if not ModuleInstaller("git@github.com:" + parsed_url.path[1:]).install():
        print("Skipping dependency {}: Already installed.".format(url))
    print("Successfully installed dependency " + url)

class ModuleInstaller:
    def __init__(self, git_url):
        self.git_url = git_url
        self.name = self.git_url.split("/")[-1].split(".git")[0]
        self.dir = SERVER_DIR + "modules/" + self.name
        self.config = {}

    def install(self):
        if os.system("git clone %s %s" % (self.git_url, self.dir)) is not 0:
            print("Couldn't clone the repository. Please check that the Git URL exists and that your SSH key is valid.")
            return False
        self.config = json.loads(utils.read_file(self.dir + "/eta.json"))
        if "dependencies" in self.config:
            for dependency_url in self.config["dependencies"]:
                get_dependency(dependency_url)
        self.fire_hook("preinstall")
        if self.config["name"] != self.name:
            self.name = self.config["name"]
            old_dir = self.dir
            self.dir = SERVER_DIR + "modules/" + self.name
            shutil.move(old_dir, self.dir)
        os.chdir(self.dir)
        os.system("npm i --only=dev")
        os.system("npm i --only=prod")
        for static_dir in self.config["dirs"]["staticFiles"]:
            js_dir = "%s/%s/js" % (self.dir, static_dir)
            if not os.path.isdir(js_dir):
                continue
            os.chdir(js_dir)
            print(os.getcwd())
            os.system("npm i --only=dev")
            if os.path.exists(js_dir + "/typings.json"):
                os.system("node " + SERVER_DIR + "node_modules/typings/dist/bin.js i")
            os.chdir(self.dir)
        self.fire_hook("postinstall")
        os.chdir(SERVER_DIR)
        os.system("npm run generate-ci")
        os.system("npm run compile-client")
        return True

    def fire_hook(self, hook_name):
        if "hooks" not in self.config:
            return
        if hook_name not in self.config["hooks"]:
            return
        for hook in self.config["hooks"][hook_name]:
            if "cwd" in hook:
                os.chdir(self.dir + "/" + hook["cwd"])
            os.system(hook["exec"])
            os.chdir(self.dir)

def main():
    git_url = sys.argv[1] if len(sys.argv) >= 2 else utils.get_input("Git URL: ")
    ModuleInstaller(git_url).install()

if __name__ == "__main__":
    main()

File: scripts/get-module.ts
import * as fs from "fs-extra";
import * as utils from "./utils";
import HelperFS from "../helpers/fs";

const SERVER_DIR: string = utils.getServerDir();

class ModuleInstaller {
    private url: string;
    private name: string;
    private path: string;
    private config: {
        dirs: {
            controllers: string[];
            models: string[];
            staticFiles: string[];
            views: string[];
            lifecycleHandlers: string[];
            requestTransformers: string[];
        };
        css: {[key: string]: string};
        name: string;
        redirects: {[key: string]: string};
        rootDir: string;
        dependencies: string[];
        hooks: {[key: string]: {cwd: string, exec: string}[]};
    };

    public constructor(url: string) {
        this.url = url;
        this.name = this.url.split("/").slice(-1)[0].replace(/\.git/g, "");
        this.path = SERVER_DIR + "/modules/" + this.name;
    }

    public async install(isDependency = true): Promise<boolean> {
        try {
            await utils.exec(`git clone ${this.url} ${this.path}`, { cwd: SERVER_DIR });
        } catch (err) {
            if (!isDependency) {
                console.error("Couldn't clone the repository. Please check that the Git URL exists and that your SSH key is valid.");
            }
            return false;
        }
        this.config = JSON.parse(await fs.readFile(this.path + "/eta.json", "utf-8"));
        if (this.config.dependencies) {
            for (const url of this.config.dependencies) {
                // dependency installation
                if (url.split("/").length !== 2) {
                    console.error("Please format dependency URLs as: username/repository (only Github repositories)");
                    console.error("Couldn't get dependency: " + url);
                    continue;
                }
                console.log("Installing dependency... (" + url + ")");
                const installer: ModuleInstaller = new ModuleInstaller("git@github.com:" + url);
                try {
                    if (await installer.install()) {
                        console.log("Successfully installed dependency: " + url);
                    } else {
                        console.log("Skipping dependency " + url + ": Already installed.");
                    }
                } catch (err) {
                    console.error("Couldn't install dependency: " + url, err);
                }
            }
        }
        await this.fireHook("preinstall");
        if (this.config.name !== this.name) {
            this.name = this.config.name;
            await fs.move(this.path, SERVER_DIR + "/modules/" + this.name);
            this.path = SERVER_DIR + "/modules/" + this.name;
        }
        console.log("\tInstalling NPM modules...");
        await utils.exec("npm i --only=dev", { cwd: this.path });
        await utils.exec("npm i --only=prod", { cwd: this.path });
        console.log("\tSetting up client-side JS...");
        for (const staticPath of this.config.dirs.staticFiles) {
            const jsPath = `${this.path}/${staticPath}/js`;
            if (!(await HelperFS.exists(jsPath))) {
                continue;
            }
            if (await HelperFS.exists(jsPath + "/typings.json")) {
                await utils.exec("node " + SERVER_DIR + "node_modules/typings/dist/bin.js i", { cwd: jsPath });
            }
        }
        console.log("\tCompiling server-side JS...");
        await utils.exec("npm run generate-ci", { cwd: SERVER_DIR });
        console.log("\tCompiling client-side JS...");
        await utils.exec("npm run compile-client", { cwd: SERVER_DIR });
        return true;
    }

    private async fireHook(name: string): Promise<void> {
        if (!this.config.hooks || !this.config.hooks[name]) {
            return;
        }
        for (const hook of this.config.hooks[name]) {
            const options: any = {};
            if (hook.cwd) {
                options.cwd = this.path + "/" + hook.cwd;
            }
            await utils.exec(hook.exec, options);
        }
    }
}

async function main(): Promise<void> {
    let url: string = process.argv[2];
    if (!url) {
        console.log("Usage: npm run get-module -- <git-url>");
    }
    if (!url.startsWith("git@") && !url.startsWith("https://")) {
        url = "git@github.com:" + url;
    }
    const installer: ModuleInstaller = new ModuleInstaller(url);
    await installer.install(false);
}

if (!module.parent) {
    main().then(() => { })
    .catch(err => {
        console.error(err);
    });
}
